wgpu-core/src/error.rs

```diff
@@ -0,0 +1,321 @@
+use core::fmt;
+use std::error::Error;
+
+use crate::{
+    gfx_select,
+    hub::{Global, IdentityManagerFactory},
+};
+
+pub trait AsDisplay {
+    fn as_display(&self) -> &dyn fmt::Display;
+}
+
+impl<T: fmt::Display> AsDisplay for T {
+    fn as_display(&self) -> &dyn fmt::Display {
+        self
+    }
+}
+
+pub trait PrettyError: Error {
+    fn fmt_pretty(&self, _global: &Global<IdentityManagerFactory>) -> String {
+        format_error_line(self.as_display())
+    }
+}
+
+impl PrettyError for super::command::RenderCommandError {
+    fn fmt_pretty(&self, global: &Global<IdentityManagerFactory>) -> String {
+        let mut ret = format_error_line(self);
+        match *self {
+            Self::InvalidBindGroup(id) => {
+                let name = gfx_select!(id => global.bind_group_label(id));
+                ret.push_str(&format_label_line("bind group", &name));
+            }
+            Self::InvalidPipeline(id) => {
+                let name = gfx_select!(id => global.render_pipeline_label(id));
+                ret.push_str(&format_label_line("render pipeline", &name));
+            }
+            Self::Buffer(id, ..) | Self::DestroyedBuffer(id) => {
+                let name = gfx_select!(id => global.buffer_label(id));
+                ret.push_str(&format_label_line("buffer", &name));
+            }
+            _ => {}
+        };
+        ret
+    }
+}
+impl PrettyError for crate::binding_model::CreateBindGroupError {
+    fn fmt_pretty(&self, global: &Global<IdentityManagerFactory>) -> String {
+        let mut ret = format_error_line(self);
+        match *self {
+            Self::InvalidBuffer(id) => {
+                let name = crate::gfx_select!(id => global.buffer_label(id));
+                ret.push_str(&format_label_line("buffer", &name));
+            }
+            Self::InvalidTextureView(id) => {
+                let name = crate::gfx_select!(id => global.texture_view_label(id));
+                ret.push_str(&format_label_line("texture view", &name));
+            }
+            Self::InvalidSampler(id) => {
+                let name = crate::gfx_select!(id => global.sampler_label(id));
+                ret.push_str(&format_label_line("sampler", &name));
+            }
+            _ => {}
+        };
+        ret
+    }
+}
+
+impl PrettyError for crate::binding_model::CreatePipelineLayoutError {
```

Comment:
Would it make sense to move each implementation into the module that own the type (i.e. `CreatePipelineLayoutError` in this case)? This way, if the error gets more variants, it's easier to see the associated code.

Reply:
Ah yeah, true, now that that is possible and makes sense.
Does the `PrettyError` name still make sense if it is used more widely?

I'm somehow still hoping for a way to implement the formatting with the standard `Display` trait, but I guess there is still no way of wiggling the Global in there to resolve the labels. Does pythonesques rewrite affect this, would the label be accessible just from the Id? That would simplify the whole thing, if we could just format the errors like any other errors.

Reply:
cc @pythonesque

> Does the PrettyError name still make sense if it is used more widely?

seems fine, unless you have ideas for a better name ;)